src/main/java/io/supertokens/webserver/api/bulkimport/AddBulkImportUsers.java

```diff
@@ -0,0 +1,170 @@
+            errorResponseJson.add("users", errorsJson);
+            throw new ServletException(new WebserverAPI.BadRequestException(errorResponseJson.toString()));
+        }
+
+        try {
+            BulkImport.addUsers(appIdentifierWithStorage, usersToAdd);
+        } catch (TenantOrAppNotFoundException | StorageQueryException e) {
+            throw new ServletException(e);
+        }
+
+        JsonObject result = new JsonObject();
+        result.addProperty("status", "OK");
+        super.sendJsonResponse(200, result, resp);
+    }
+
+    private void validateTenantId(AppIdentifier appIdentifier, String tenantId, String recipeId)
+            throws InvalidBulkImportDataException, ServletException {
+        if (tenantId == null || tenantId.equals(TenantIdentifier.DEFAULT_TENANT_ID)) {
+            return;
+        }
+
+        try {
+            if (Arrays.stream(FeatureFlag.getInstance(main, appIdentifier).getEnabledFeatures())
+                    .noneMatch(t -> t == EE_FEATURES.MULTI_TENANCY)) {
+                throw new InvalidBulkImportDataException(new ArrayList<>(
+                        Arrays.asList("Multitenancy must be enabled before importing users to a different tenant.")));
+            }
+        } catch (TenantOrAppNotFoundException | StorageQueryException e) {
+            throw new ServletException(e);
+        }
+
+        TenantConfig[] allTenantConfigs = Multitenancy
+                .getAllTenantsForApp(appIdentifier, main);
+        ArrayList<String> validTenantIds = new ArrayList<>();
+        Arrays.stream(allTenantConfigs)
+                .forEach(tenantConfig -> validTenantIds.add(tenantConfig.tenantIdentifier.getTenantId()));
+
+        if (!validTenantIds.contains(tenantId)) {
+            throw new InvalidBulkImportDataException(
+                    new ArrayList<>(Arrays.asList("Invalid tenantId: " + tenantId + " for " + recipeId + " recipe.")));
+        }
+    }
+
+    private void validatePasswordHashingAlgorithm(AppIdentifier appIdentifier, EmailPasswordLoginMethod emailPasswordLoginMethod) throws InvalidBulkImportDataException, ServletException {
+        try {
+            CoreConfig.PASSWORD_HASHING_ALG passwordHashingAlgorithm = CoreConfig.PASSWORD_HASHING_ALG.valueOf(emailPasswordLoginMethod.hashingAlgorithm);
+            PasswordHashingUtils.assertSuperTokensSupportInputPasswordHashFormat(appIdentifier, main, emailPasswordLoginMethod.passwordHash, passwordHashingAlgorithm);
+        } catch (UnsupportedPasswordHashingFormatException | TenantOrAppNotFoundException e) { 
+            throw new InvalidBulkImportDataException(new ArrayList<>(Arrays.asList(e.getMessage())));
+        } catch (IllegalArgumentException e) {
+            throw new InvalidBulkImportDataException(new ArrayList<>(Arrays.asList("Invalid hashingAlgorithm for emailpassword recipe. Pass one of bcrypt, argon2 or, firebase_scrypt!")));
+        }
```

Comment:
TenantOrAppNotFoundException -> should be propagated to the calling function.

Reply:
done